HTTP method support in MappedInterceptor

SRIRAM9487 opened the pull request:
# Support HTTP method-restricted interceptors in `InterceptorRegistration`

This PR implements [#35272](https://github.com/spring-projects/spring-framework/issues/35272), adding support for restricting `HandlerInterceptor` registration to specific HTTP methods.

## Motivation

Spring MVC currently allows interceptors to be applied based on path patterns, but not HTTP methods. When method-based filtering is needed, it must be manually implemented within the interceptor itself:

```java
@Override
public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
    if (!"POST".equals(request.getMethod())) {
        return true;
    }
    // Logic here
    return true;
}
```

This approach clutters interceptor logic with method checks and duplicates similar logic across interceptors when needed. Centralizing HTTP method filtering in the registration API provides a cleaner and more declarative configuration style.

## Solution

This PR introduces a new method to `InterceptorRegistration`:

```java
public InterceptorRegistration allowedMethods(HttpMethod... methods)
```

This allows restricting the interceptor to specific HTTP methods like `GET`, `POST`, etc. Under the hood, the framework wraps the given `HandlerInterceptor` in a new `MethodInterceptor` that delegates calls only when the request's HTTP method matches one of the allowed methods.

```java
registry.addInterceptor(new MyInterceptor())
        .addPathPatterns("/api/**")
        .allowedMethods(HttpMethod.POST, HttpMethod.PUT);
```

The interceptor will now be invoked **only** for POST and PUT requests matching `/api/**`.

## Implementation Details

- `InterceptorRegistration` stores allowed HTTP methods as a `Set<String>`.
- If any methods are configured, it wraps the given interceptor using the new `MethodInterceptor`.
- `MethodInterceptor` checks the request's method before delegating to the original interceptor.
- The implementation is fully backward compatible; if no methods are provided, all HTTP methods are allowed (default behavior).

## Tests

A dedicated test class `MethodInterceptorTests` is included to validate:

- Delegation occurs only for allowed HTTP methods.
- Interceptor is skipped for disallowed methods.
- All interceptor lifecycle methods (`preHandle`, `postHandle`, `afterCompletion`) are covered.

## Example

```java
@Configuration
public class WebConfig implements WebMvcConfigurer {

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(new AuditInterceptor())
                .addPathPatterns("/admin/**")
                .allowedMethods(HttpMethod.GET);
    }
}
```

In this example, `AuditInterceptor` will apply only for `GET` requests under `/admin/**`.

## Compatibility

- No breaking changes introduced.
- The feature is opt-in.
- Reuses existing Spring MVC components (`MappedInterceptor`, `HandlerInterceptor`) without exposing new public types.

SRIRAM9487 commented:
Thanks for the feedback on #35272. Understood that this feature is not currently planned.

Would it help if I maintained this functionality as an extension or community module?  
I still believe it solves a real-world use cases ,  
but I completely respect your direction here.

Thanks again for your time and consideration!

SRIRAM9487 commented:
Thanks for the clarification, @bclozel.
Looking forward to your review once it's in the queue. Let me know if anything needs to be adjusted in the meantime!

rissay2015 commented:
Until then (the review by those grannies) : 
   ->You can work on exclude methodes  : applying that interceptor to all except this methodes ....
   -> feels like we must fall back to manual logic inside the interceptor for anything more complex.
Otherwise that was impressive young man.

rstoyanchev reviewed (changes requested):
Aside from the comments below, generally the request seems reasonable with the following implementation related points:

- The existing `MappedInterceptor` could be enhanced instead to support mapping by HTTP method.
- The terminology `allowedMethods` could be just `httpMethods`, or `include/excludeHttpMethods` for consistency with `include/excludePatterns`.

SRIRAM9487 committed:
Add HTTP method-based filtering support to MappedInterceptor and InterceptorRegistration

- Extended MappedInterceptor to support include/exclude HTTP methods for request matching
- Added corresponding constructors for HandlerInterceptor and WebRequestInterceptor
- Updated InterceptorRegistration with fluent methods to configure HTTP method filters
- Ensured backward compatibility by keeping existing constructors and methods intact
- Improved matches() logic to combine URL pattern and HTTP method conditions

This enhancement enables finer control over interceptor application based on HTTP methods,
aligning with modern Spring 7.x practices.

Signed-off-by: SRIRAM9487 <sriram9487tk@gmail.com>

SRIRAM9487 commented on spring-aop/.factorypath:
```diff
@@ -0,0 +1,25 @@
+    <factorypathentry kind="EXTJAR" id="/home/sriram/.gradle/caches/modules-2/files-2.1/org.pcollections/pcollections/4.0.1/59f3bf5fb28c5f5386804dcf129267416b75d7c/pcollections-4.0.1.jar" enabled="true" runInBatchMode="false"/>
+    <factorypathentry kind="EXTJAR" id="/home/sriram/.gradle/caches/modules-2/files-2.1/com.uber.nullaway/nullaway/0.12.7/1a813b7d156768204b0c8d52ba0632a8db6fbe12/nullaway-0.12.7.jar" enabled="true" runInBatchMode="false"/>
+    <factorypathentry kind="EXTJAR" id="/home/sriram/.gradle/caches/modules-2/files-2.1/com.google.guava/failureaccess/1.0.3/aeaffd00d57023a2c947393ed251f0354f0985fc/failureaccess-1.0.3.jar" enabled="true" runInBatchMode="false"/>
+</factorypath>
```
I've removed the unintended .factorypath entries from the pull request.

SRIRAM9487 commented on spring-webmvc/src/main/java/org/springframework/web/servlet/config/annotation/InterceptorRegistration.java:
```diff
@@ -75,18 +79,19 @@ public InterceptorRegistration addPathPatterns(String... patterns) {
 	public InterceptorRegistration addPathPatterns(List<String> patterns) {
-		this.includePatterns = (this.includePatterns != null ?
-				this.includePatterns : new ArrayList<>(patterns.size()));
+		this.includePatterns = (this.includePatterns != null ? this.includePatterns : new ArrayList<>(patterns.size()));
```
I've reverted the unintended code style changes in InterceptorRegistration.java to match the existing formatting in the codebase.

SRIRAM9487 committed:
unit tests for HTTP method filtering in MappedInterceptor

- Verify inclusion of single and multiple HTTP methods
- Verify exclusion of single and multiple HTTP methods
- Test combined include and exclude HTTP method filters
- Confirm correct matching behavior with URL path patterns
- Validate interceptor lifecycle method delegation (preHandle, postHandle, afterCompletion)
- All tests pass successfully, ensuring robust coverage of method filtering logic

Signed-off-by: SRIRAM9487 <sriram9487tk@gmail.com>

SRIRAM9487 commented:
@rstoyanchev Thank you for reviewing and integrating this feature into Spring Framework. I appreciate the refinements in the final implementation,. I initially thought it might be merged directly, but I understand it’s a relatively small feature. Looking forward to contributing again in the future.
